### PR TITLE
Data disk support for Cloudstack

### PR DESCRIPTION
**Pull request by jeroendekorte**
The following feature for the Cloudstack plugin enables the possibility to have multiple disks on a virtual machine. From the blueprint add a sperate volume_node and attach it with a contained_in relationship. Rollback is configured to default, detach the volume from the vm and destroy the disk if the purge attribute is set to true in the blueprint.

### CODE REVIEW

File: cloudstack_plugin/volume.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    COMMON_RUNTIME_PROPERTIES_KEYS
+
+)
+from cloudstack_plugin.virtual_machine import get_vm_by_id
+
+VOLUME_CLOUDSTACK_TYPE = 'volume'
+
+RUNTIME_PROPERTIES_KEYS = COMMON_RUNTIME_PROPERTIES_KEYS
+
+__author__ = 'jedeko'
```
Comment (idanmo) [inline cloudstack_plugin/volume.py]: Hi!
Please don't include __author__ statements in source files.
Thanks!

File: cloudstack_plugin/volume.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+
+@operation
+def attach_volume(ctx, **kwargs):
```
Comment (idanmo) [inline cloudstack_plugin/volume.py]: It would be nice if you could import ctx instead of passing it as an argument:
from cloudify import ctx

File: cloudstack_plugin/volume.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        ctx.logger.warn('Volume {0} may not have been detached: {1}'
+                        .format(volume, str(e)))
+        pass
+
+    # Expunge the volume if true in blueprint configuration
+    volume_expunge = ctx.source.node.properties['expunge']
+
+    if volume_expunge:
+        try:
+            cloud_driver.destroy_volume(volume=volume)
```
Comment (idanmo) [inline cloudstack_plugin/volume.py]: Perhaps add a log message saying there's an attempt to destroy the volume?

File: plugin.yaml [hunk 4]
```diff
@@ -207,6 +223,17 @@ relationships:
 #                 - establish: cloudstack.neutron_plugin.router.connect_subnet
 #                 - unlink: cloudstack.neutron_plugin.router.disconnect_subnet
 
+    cloudify.cloudstack.volume_contained_in_vm:
+            derived_from: cloudify.relationships.contained_in
+            source_interfaces:
+                cloudify.interfaces.relationship_lifecycle:
+                    establish:
+                        implementation: cloudstack.cloudstack_plugin.volume.attach_volume
```
Comment (idanmo) [inline plugin.yaml]: You can map the operations this way:
establish: cloudstack.cloudstack_plugin.volume.attach_volume
etc...

**Comment by jeroendekorte** [comment]
Hi Idan,

Thanks for the quick reply and comments. Still in a learning phase so all the help is welcome. I've changed the code as requested. 

Regards,
Jeroen

File: cloudstack_plugin/volume.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+@operation
+def detach_volume(**kwargs):
+    """ Detaches a volume and delete if expunge is requested.
+    """
+
+    cloud_driver = get_cloud_driver(ctx)
+
+    volume_id = ctx.source.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY]
+    volume = get_volume_by_id(cloud_driver, volume_id)
```
Comment (idanmo) [inline cloudstack_plugin/volume.py]: get_volume_by_id can return None so perhaps it would be best to raise a NonRecoverableError if the volume doesn't exist in this method.

File: README.md [hunk 6]
```diff
@@ -19,7 +19,7 @@ Very condensed HOWTO, more to follow:
 * pip install https://github.com/cloudify-cosmo/cloudify-rest-client/archive/master.zip
 * pip install https://github.com/cloudify-cosmo/cloudify-plugins-common/archive/master.zip
 * pip install https://github.com/cloudify-cosmo/cloudify-script-plugin/archive/master.zip
-*pip install https://github.com/cloudify-cosmo/cloudify-cli/archive/master.zip
+* pip install https://github.com/cloudify-cosmo/cloudify-cli/archive/master.zip
 * git clone https://github.com/schubergphilis/cloudify-manager-blueprints.git
 * git clone https://github.com/schubergphilis/cloudify-nodecellar-example.git
 * cp cloudify-manager-blueprints/cloudstack/inputs.json.template cloudify-config.json
```
Comment (iliapolo) [inline README.md]: note that the input files in the manager-blueprints repo were changed to be yaml files.

please fix the readme accordingly.

File: plugin.yaml [hunk 7]
```diff
@@ -207,6 +223,13 @@ relationships:
 #                 - establish: cloudstack.neutron_plugin.router.connect_subnet
 #                 - unlink: cloudstack.neutron_plugin.router.disconnect_subnet
 
+    cloudify.cloudstack.volume_contained_in_vm:
```
Comment (iliapolo) [inline plugin.yaml]: volume should be attached to vm, not contained in.

have a look at the openstack example:
https://github.com/cloudify-cosmo/cloudify-openstack-plugin/blob/master/plugin.yaml#L297

File: README.md (hunk 6 above)
Comment (idanmo) [inline README.md]: Just to clarify, both YAML and JSON formats are supported.

File: plugin.yaml [hunk 8]
```diff
@@ -147,6 +147,22 @@ node_types:
             cloudstack_config:
                 default: {}
 
+    cloudify.cloudstack.nodes.Volume:
+            derived_from: cloudify.nodes.Volume
+            properties:
+                name:
+                    default: ''
```
Comment (iliapolo) [inline plugin.yaml]: is this valid? maybe this should not have a default value, so there will be validation when parsing the dsl.

same goes for size

File: cloudstack_plugin/volume.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+    cloud_driver = get_cloud_driver(ctx)
+
+    vm_id = ctx.target.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY]
+    vm = get_vm_by_id(ctx, cloud_driver, vm_id)
+
+    volume_name = ctx.source.node.properties['name']
+    volume_size = ctx.source.node.properties['size']
+
+    volume = cloud_driver.create_volume(name=volume_name,
```
Comment (iliapolo) [inline cloudstack_plugin/volume.py]: why is 'attach_volume' creating the volume?

creation and deletion of volumes should be part of the volume lifecycle, not in the relationship

File: cloudstack_plugin/volume.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+    # Expunge the volume if true in blueprint configuration
+    volume_expunge = ctx.source.node.properties['expunge']
+
+    if volume_expunge:
+        try:
+            ctx.logger.info('Trying to destroy volume {0}'
+                            .format(volume))
+
+            cloud_driver.destroy_volume(volume=volume)
```
Comment (iliapolo) [inline cloudstack_plugin/volume.py]: see comment for attach

File: cloudstack_plugin/volume.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+    volume = cloud_driver.create_volume(name=volume_name,
+                                        size=volume_size)
+
+    cloud_driver.attach_volume(node=vm,
+                               volume=volume)
+
+    ctx.source.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY] = volume.id
+    ctx.source.instance.runtime_properties[CLOUDSTACK_TYPE_PROPERTY] = \
+        VOLUME_CLOUDSTACK_TYPE
```
Comment (iliapolo) [inline cloudstack_plugin/volume.py]: is there any way of knowing what the device name of the attached volume? i mean, vdb, vdc, vda...
we need this information to use file systems with this volume.

File: cloudstack_plugin/volume.py (hunk 9 above)
Comment (jeroendekorte) [inline cloudstack_plugin/volume.py]: If I understand correctly, there should be seperate functions for creating, deleting and attaching a volume? Is that what you mean by this?

Comment (iliapolo) [inline cloudstack_plugin/volume.py]: yes, and the 'create' and 'delete' functions should be mapped to the volume lifecycle operations.

File: cloudstack_plugin/volume.py (hunk 11 above)
Comment (boul) [inline cloudstack_plugin/volume.py]: Device name is a bit harder as it is very much depended on the underlying hypervisor which is normally unknown to a user of cloudstack. e.g. /dev/xvda for xen in paravirt mode but can also become /dev/sda or /dev/vda depending on hypervisor and HVM vs paravirt. 
Attach volume does return a deviceid, so let's store that anyway and then we need some extra magic to figure out the actual device name when extending the actual filesystem. Does that make sense?

Comment (iliapolo) [inline cloudstack_plugin/volume.py]: not sure i follow. what is deviceid 1 and deviceid 2? why are there two of them?

there are two options for the built in file system type to work:
1. the volume node contains a 'device_name' property that is the actual device name. This is good in case the API accepts such an argument and can guaranty that this device name is correct. since, like you said, this highly depends on the hypervisor, most cloud providers cannot enforce it.
2. option two, is that the attach operation will set a runtime property on the source called 'device_name'.

so given that, is there any way of achieving it with cloudstack? 
for example, openstack has an option to pass 'auto' as the device name, and later you can query the attachement and it will return the actual device name.

sorry for the long comment, just wanted to explain the missing parts.

Comment (boul) [inline cloudstack_plugin/volume.py]: Sorry for being unclear. 

when doing an attach_volume call to cloudstack it only returns a device_id, where 1 maps /dev/xvdb or vdb or sdb (depending on hypervisor)(and 2 maps to /dev/xvdc etc) so it seems logical to at least store this deviceid as a runtime property. There is no way currently, to query cloudstack for the actual device name. So storing an assumed device-name such as /dev/xvdb which can also be /dev/vdb depending on hypervisor and such seems confusing. 

It seems better to me to find out the actual device name upon filesystem creation/expansion knowing that device_id=2 will be either /dev/xvdc OR /dev/vdc OR /dev/sdc. 

see also: https://cloudstack.apache.org/docs/api/apidocs-4.1/root_admin/attachVolume.html

**Comment by jeroendekorte** [comment]
As explained by Eli, I have implemented the "create" and "delete" functions as a volume lifecycle operation and implemented the attach and detach functions in de virtual_machine operations. Tested everything successfully.

File: cloudstack_plugin/virtual_machine.py [hunk 12]
```diff
@@ -602,6 +606,61 @@ def disconnect_floating_ip(ctx, **kwargs):
...
+
+    server = get_vm_by_id(ctx, cloud_driver, server_id)
+    volume = get_volume_by_id(cloud_driver, volume_id)
+
+    if server is None:
+        raise NonRecoverableError('Server with id {0} not found'
+                                  .format(server_id))
+
+    if volume is None:
+        if volume is None:
```
Comment (iliapolo) [inline cloudstack_plugin/virtual_machine.py]: looks like a copy paste mistake...duplicate 'if volume is None'

File: cloudstack_plugin/virtual_machine.py [hunk 13]
```diff
@@ -602,6 +606,61 @@ def disconnect_floating_ip(ctx, **kwargs):
...
+            raise NonRecoverableError('Volume with id {0} not found'
+                                      .format(volume_id))
+
+    cloud_driver.attach_volume(node=server,
+                               volume=volume)
```
Comment (iliapolo) [inline cloudstack_plugin/virtual_machine.py]: what about the device name? is it possible to set a runtime property on the source specifying the device name?

File: cloudstack_plugin/virtual_machine.py [hunk 14]
```diff
@@ -602,6 +606,61 @@ def disconnect_floating_ip(ctx, **kwargs):
...
+
+    # Detach the volume
+    try:
+        cloud_driver.detach_volume(volume=volume)
+    except Exception as e:
+        ctx.logger.warn('Volume {0} may not have been detached: {1}'
+                        .format(volume, str(e)))
+        pass
+
+    # # Expunge the volume if true in blueprint configuration
```
Comment (iliapolo) [inline cloudstack_plugin/virtual_machine.py]: delete this code for now.

commented code should be avoided...

File: plugin.yaml [hunk 15]
```diff
@@ -147,6 +146,21 @@ node_types:
             cloudstack_config:
                 default: {}
 
+    cloudify.cloudstack.nodes.Volume:
+            derived_from: cloudify.nodes.Volume
+            properties:
+                volume: {}
+                use_external_resource:
```
Comment (iliapolo) [inline plugin.yaml]: where are  you using this property? 
'create' should read this and not do anything if its set to true.

File: cloudstack_plugin/virtual_machine.py (hunk 14 above)
Comment (jeroendekorte) [inline cloudstack_plugin/virtual_machine.py]: Should be implemented, I will in the next commit.

File: plugin.yaml (hunk 15 above)
Comment (jeroendekorte) [inline plugin.yaml]: Which part of this are you talking about?

**Comment by jeroendekorte** [comment]
Hi,

Made some changes as requested. The only thing I cannot implement is the volume name when created. This (volume name) is not something Cloudstack is returning, only a id.

**Comment by jeroendekorte** [comment]
Guys,

Did not get any response since my last comment. Could you be so kind to have a look at this pull request?

Regards,
Jeroen

**Comment by iliapolo** [comment]
Hey, having a look right now

File: cloudstack_plugin/virtual_machine.py [hunk 16]
```diff
@@ -602,6 +606,62 @@ def disconnect_floating_ip(ctx, **kwargs):
...
+
+    # Detach the volume
+    try:
+        cloud_driver.detach_volume(volume=volume)
+    except Exception as e:
+        ctx.logger.warn('Volume {0} may not have been detached: {1}'
+                        .format(volume, str(e)))
+        pass
+
+    # Expunge the volume if true in blueprint configuration
```
Comment (iliapolo) [inline cloudstack_plugin/virtual_machine.py]: why is this part needed here? its already implemented in delete...
detach should just detach, not delete

File: cloudstack_plugin/volume.py [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+    cloud_driver = get_cloud_driver(ctx)
+
+    volume = {}
+
+    ctx.logger.debug('reading volume attributes.')
+    volume.update(copy.deepcopy(ctx.node.properties['volume']))
+
+    volume_name = volume['name']
+    volume_size = volume['size']
+
```
Comment (iliapolo) [inline cloudstack_plugin/volume.py]: if 'name' or 'size' are not part of the volume, a KeyError will be raised, which will be translated into a recoverable error, which is not what we want. 
check if these keys exist and if not than raise a NonRecoverableError

File: cloudstack_plugin/volume.py [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+    cloud_driver = get_cloud_driver(ctx)
+
+    volume_id = ctx.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY]
+    volume = get_volume_by_id(cloud_driver, volume_id)
+
+    # Expunge the volume if true in blueprint configuration
+    volume_properties = ctx.node.properties['volume']
+
+    volume_expunge = volume_properties['expunge']
```
Comment (iliapolo) [inline cloudstack_plugin/volume.py]: expunge may not be in volume. default value should be true for now until we figure out how exactly we want this scenarios to handled

File: plugin.yaml (hunk 15 above)
Comment (iliapolo) [inline plugin.yaml]: the use_external_resource property.
i dont any use of it in the code...

File: cloudstack_plugin/volume.py (hunk 18 above)
Comment (jeroendekorte) [inline cloudstack_plugin/volume.py]: Thanks Eli for the comments, I've updated the pull request

File: cloudstack_plugin/virtual_machine.py [hunk 19]
```diff
@@ -602,6 +606,53 @@ def disconnect_floating_ip(ctx, **kwargs):
...
+    volume_id = ctx.target.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY]
+    volume = get_volume_by_id(cloud_driver, volume_id)
+
+    if volume is None:
+        raise NonRecoverableError('Volume with id {0} not found'
+                                  .format(volume_id))
+
+    try:
+        cloud_driver.detach_volume(volume=volume)
+    except Exception as e:
```
Comment (iliapolo) [inline cloudstack_plugin/virtual_machine.py]: this is probably not needed anymore since its the last operation of the method, let if raise the exception as is...

File: cloudstack_plugin/volume.py [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+                                  .format(volume_id))
+
+    try:
+        ctx.logger.info('Trying to destroy volume {0}'
+                        .format(volume))
+
+        cloud_driver.destroy_volume(volume=volume)
+
+    except Exception as e:
+        ctx.logger.warn('Volume {0} may not have been destroyed: {1}'
```
Comment (iliapolo) [inline cloudstack_plugin/volume.py]: same comment from above

File: plugin.yaml (hunk 15 above)
Comment (iliapolo) [inline plugin.yaml]: what about this then? if we dont currently support using external volumes in cloudstack than this property should not be here. right?

File: cloudstack_plugin/volume.py (hunk 20 above)
Comment (jeroendekorte) [inline cloudstack_plugin/volume.py]: Hi Eli,

I've updated the methods now. 

Thanks!

**Comment by jeroendekorte** [comment]
Hi Guys,

Any update on this? Thanks!

Regards,
Jeroen

**Comment by iliapolo** [comment]
Hi, i left a comment regarding the 'use_external_resource' property that you defined in the volume type, but i dont see its usage in the code...

**Comment by jeroendekorte** [comment]
Hi Eli,

Missed that one, removed it now. Thanks

**Comment by iliapolo** [comment]
Hmm, i actually meant that this should be implemented for volume :)

i mean, cloudstack plugin does offer this use case for other resources, so i think volume should be no different...

**Comment by jeroendekorte** [comment]
Hi Eli,

Please review if this is correctly done. I've tested it and it works fine. Not sure though if this is the way to do it. Thanks!

File: cloudstack_plugin/virtual_machine.py (hunk 19 above)
Comment (iliapolo) [inline cloudstack_plugin/virtual_machine.py]: i dont see a change here...please remove the 'except' clause

File: cloudstack_plugin/volume.py (hunk 20 above)
Comment (iliapolo) [inline cloudstack_plugin/volume.py]: where is the update? i meant that the 'except' clause should be removed. same thing for the detach

Comment (iliapolo) [inline cloudstack_plugin/volume.py]: also, delete should consider the 'use_external_resource' property as well. if it is set to true than we dont delete the volume

**Comment by iliapolo** [comment]
Hi.

please see my new comments....

**Comment by iliapolo** [comment]
Hi, please see my new comments...

**Comment by av1nashs1ngh** [comment]
Hi, Volume now respects the use_external_resource property, can you guys have a look again? Thanks!

File: cloudstack_plugin/volume.py [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+    """
+
+    cloud_driver = get_cloud_driver(ctx)
+
+    volume = {}
+
+    ctx.logger.debug('reading volume attributes.')
+    volume.update(copy.deepcopy(ctx.node.properties['volume']))
+
+    if 'name' in volume:
```
Comment (iliapolo) [inline cloudstack_plugin/volume.py]: what happens if use_external_resource is set to True, and a resource id is supplied, but the name and size are left blank. in this case it will raise an error, while it shouldnt since we are not really going to create the volume...

File: cloudstack_plugin/volume.py [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+
+        ctx.logger.info('Using existing volume: {0}'.
+                        format(volume_name))
+
+        volume = get_volume_by_name(cloud_driver, volume_name)
+
+        ctx.instance.runtime_properties[CLOUDSTACK_ID_PROPERTY] = volume.id
+        ctx.instance.runtime_properties[CLOUDSTACK_NAME_PROPERTY] = volume.name
+        ctx.instance.runtime_properties[CLOUDSTACK_TYPE_PROPERTY] = \
+            VOLUME_CLOUDSTACK_TYPE
```
Comment (iliapolo) [inline cloudstack_plugin/volume.py]: This implementation does not seem to follow the proper semantics.

What is should to is the following:

if use_external_resource is True, lookup the resource by the resource_id field, if resource_id is not supplied, raise an error. if the resource does not exist, raise an error. if it does exist, set the needed runtime properties and thats it.

if use_external_resource is False, try creating the volume and apply all the validations for size and name.
if a volume with that name already exists than thats ok, currently we allow creating two volumes with the same name, openstack allows it, and so should we, because the unique identifier of a resource is its id, not name.

please have deeper look at how the openstack plugin handles this.

**Comment by jeroendekorte** [comment]
Hi Eli,

Thanks for the clear comment on this. I changed it accordingly and tested it with the different settings (use_external yes and no).

Regards,
Jeroen

**Comment by iliapolo** [comment]
please resolve the conflicts...

File: cloudstack_plugin/virtual_machine.py [hunk 23]
```diff
@@ -602,6 +608,63 @@ def disconnect_floating_ip(ctx, **kwargs):
...
+
+        if volume is None:
+            raise NonRecoverableError('Volume with id {0} not found'
+                                      .format(volume_id))
+
+        cloud_driver.attach_volume(node=server,
+                                   volume=volume)
+
+    else:
+        ctx.logger.info('Volume {0} already exists and use of external '
```
Comment (iliapolo) [inline cloudstack_plugin/virtual_machine.py]: i dont understand this. why if external_id is None this means that the volume already exists? isnt it the other way around? and what does it have to do with attachement?

File: cloudstack_plugin/virtual_machine.py [hunk 24]
```diff
@@ -602,6 +608,63 @@ def disconnect_floating_ip(ctx, **kwargs):
...
+    volume = get_volume_by_id(cloud_driver, volume_id)
+
+    if not volume:
+        raise NonRecoverableError('Volume with id {0} not found'.format(
+            volume_id))
+
+    try:
+        ctx.logger.info('Detaching volume {0}'.format(volume))
+        cloud_driver.detach_volume(volume=volume)
+    except Exception as e:
```
Comment (iliapolo) [inline cloudstack_plugin/virtual_machine.py]: dont catch the exception here. just let it be raised...its an actual error.

File: cloudstack_plugin/volume.py [hunk 25]
```diff
@@ -0,0 +1,134 @@
...
+
+    elif ctx.node.properties[USE_EXTERNAL_RESOURCE_PROPERTY] is True:
+
+        if 'resource_id' in ctx.node.properties:
+            resource_id = ctx.node.properties['resource_id']
+        else:
+            raise NonRecoverableError("Resource_id is not supplied")
+
+        volume = get_volume_by_id(cloud_driver, resource_id)
+
```
Comment (iliapolo) [inline cloudstack_plugin/volume.py]: check if the volume actually exists, raise an error otherwise

File: cloudstack_plugin/volume.py [hunk 26]
```diff
@@ -0,0 +1,134 @@
...
+                                  .format(volume_id))
+    try:
+        if not ctx.node.properties[USE_EXTERNAL_RESOURCE_PROPERTY]:
+            ctx.logger.info('Trying to destroy volume {0}'.format(volume))
+            cloud_driver.destroy_volume(volume=volume)
+        else:
+            ctx.logger.info('Volume {0} does not need to be destroyed'.format(
+                volume))
+
+    except Exception as e:
```
Comment (iliapolo) [inline cloudstack_plugin/volume.py]: no need to catch an exception here. its an actual error...let it be raised.

**Comment by jeroendekorte** [comment]
Hi Eli,

Thanks again for the review, I've updated and hope this is something you can work with. 

Regards,
Jeroen

File: cloudstack_plugin/volume.py [hunk 27]
```diff
@@ -0,0 +1,129 @@
...
+
+        else:
+            raise NonRecoverableError("Volume not created")
+
+    elif ctx.node.properties[USE_EXTERNAL_RESOURCE_PROPERTY] is True:
+
+        if ctx.node.properties['resource_id']:
+            resource_id = ctx.node.properties['resource_id']
+
+            volume = get_volume_by_id(cloud_driver, resource_id)
```
Comment (iliapolo) [inline cloudstack_plugin/volume.py]: if a user specified a faulty resource_id this code would not detect it. get_volume_by_id should either raise an exception if the volume could not be found, or this code should check for None.

**Comment by jeroendekorte** [comment]
Thanks, something like this?

**Comment by iliapolo** [comment]
yes, but now for some reason there is a conflict on the merge...can you look please?

**Comment by jeroendekorte** [comment]
Hi Eli,

Should be okay now
